12-offer-encoding.md

```diff
@@ -0,0 +1,769 @@
+        * [`tu64`:`timestamp`]
+    1. type: 42 (`payment_hash`)
+    2. data:
+        * [`sha256`:`payment_hash`]
+    1. type: 44 (`relative_expiry`)
+    2. data:
+        * [`tu32`:`seconds_from_creation`]
+    1. type: 46 (`cltv`)
+    2. data:
+        * [`tu32`:`min_final_cltv_expiry`]
+    1. type: 48 (`fallbacks`)
+    2. data:
+        * [`byte`:`num`]
+        * [`num*fallback_address`:`fallbacks`]
+    1. type: 52 (`refund_signature`)
+    2. data:
+        * [`bip340sig`:`payer_signature`]
+    1. type: 56 (`replace_invoice`)
+    2. data:
+        * [`sha256`:`payment_hash`]
+    1. type: 240 (`signature`)
+    2. data:
+        * [`bip340sig`:`sig`]
+
+1. subtype: `blinded_payinfo`
+2. data:
+   * [`u32`:`fee_base_msat`]
+   * [`u32`:`fee_proportional_millionths`]
+   * [`u16`:`cltv_expiry_delta`]
+   * [`u16`:`flen`]
+   * [`flen*byte`:`features`]
+
+1. subtype: `fallback_address`
+2. data:
+   * [`byte`:`version`]
+   * [`u16`:`len`]
+   * [`len*byte`:`address`]
+
+## Requirements
+
+A writer of an invoice:
+  - MUST set `created_at` to the number of seconds since Midnight 1
+    January 1970, UTC when the offer was created.
+  - MUST set `payment_hash` to the SHA2 256-bit hash of the
+    `payment_preimage` that will be given in return for payment.
+  - MUST set (or not set) `send_invoice` the same as the offer.
+  - MUST set `offer_id` to the id of the offer.
+  - MUST specify exactly one signature TLV: `signature`.
+    - MUST set `sig` to the signature using `node_id` as described in [Signature Calculation](#signature-calculation).
+  - if the chain for the invoice is not bitcoin:
+    - MUST specify `chain` the invoice is valid for.
+  - otherwise:
+    - the bitcoin chain is implied as the first and only entry.
+  - if it has bolt11 features:
+    - MUST set `features` to the bitmap of features.
+  - if the expiry for accepting payment is not 7200 seconds after `created_at`:
+    - MUST set `relative_expiry` `seconds_from_creation` to the number of
+      seconds after `created_at` that payment of this invoice should not be attempted.
+  - if the `min_final_cltv_expiry` for the last HTLC in the route is not 18:
+    - MUST set `min_final_cltv_expiry`.
+  - if it accepts onchain payments:
+    - MAY specify `fallbacks`
+    - MUST specify `fallbacks` in order of most-preferred to least-preferred
+      if it has a preference.
+    - for the bitcoin chain, it MUST set each `fallback_address` with
+      `version` as a valid witness version and `address` as a valid witness
+      program
+  - if it is connected only by private channels:
+    - MUST include a `blinded_path` containing one or more paths to the node.
+  - otherwise:
+    - MAY include `blinded_path`.
+  - if it includes `blinded_path`:
+    - MUST specify `path` in order of most-preferred to least-preferred if
+      it has a preference.
+    - MUST include `blindedpay` with exactly one `payinfo` for
+      each `onionmsg_path` in `blinded_path`, in order.
```

Comment:
The `onionmsg_path` correspond to nodes and the `payinfo` correspond to channels, on a given route there is one less channel than there are nodes.
The number of `payinfo`should be the total number of `onionmsg_path` minus the number of `blinded_path`.

Reply:
I think it works, because blinded_path includes the first node_id, so the numbers are equal?

e.g. If the path is `A->B->C`:
1. blinded_path: first_node_id = A, num_hops = 2.
2. blinded_payinfo: needs A->B and B->C pay info.